appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+	stopCh := make(chan struct{})
+	doneCh := make(chan struct{})
+
+	var once sync.Once
+	p.cancelSyncWait = func() {
+		once.Do(func() { close(stopCh); <-doneCh })
+	}
+
+	go func() {
+		defer close(doneCh)
+
+		var status SlaveStatus
+		startTime := time.Now().UTC()
+		logger := p.Logger.New(
+			"fn", "waitForSync",
+			"sync_name", downstream.Meta["MYSQL_ID"],
+			"start_time", log15.Lazy{func() time.Time { return startTime }},
+			"log_file", log15.Lazy{func() string { return status.MasterLogFile }},
+			"log_pos", log15.Lazy{func() int64 { return status.ReadMasterLogPos }},
+		)
+
+		logger.Info("waiting for downstream replication to catch up")
+		defer logger.Info("finished waiting for downstream replication")
+
+		for {
+			// Check if "wait sync" has been canceled.
+			select {
+			case <-stopCh:
+				logger.Debug("canceled, stopping")
+				return
+			default:
+			}
+
+			// Read local master status.
+			ms, err := p.masterStatus("127.0.0.1:" + p.Port)
+			if err != nil {
+				logger.Error("error reading master status", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+
+			// Read downstream slave status.
+			ss, err := p.slaveStatus(downstream.Addr)
+			if err != nil {
+				logger.Error("error reading slave status", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+
+			elapsedTime := time.Since(startTime)
+			logger := logger.New(
+				"master_log_file", ms.File, "master_log_pos", ms.Position,
+				"slave_log_file", ss.MasterLogFile, "slave_log_pos", ss.ReadMasterLogPos,
+				"elapsed", elapsedTime,
+			)
+
+			// Update the slave status and reset start time if replication has progressed.
+			if cmp := ComparePosition(status.MasterLogFile, status.ReadMasterLogPos, ss.MasterLogFile, ss.ReadMasterLogPos); status.IsZero() || cmp == -1 {
+				logger.Debug("slave status progressing, resetting start time")
+				startTime = time.Now().UTC()
+				status = ss
+			}
+
+			logger.Info("downstream status", "master_file", ms.File, "master_pos", ms.Position, "slave_file", ss.MasterLogFile, "slave_pos", ss.ReadMasterLogPos)
+			if ms.File == ss.MasterLogFile && ms.Position == ss.ReadMasterLogPos {
+				logger.Info("downstream caught up")
+				p.syncedDownstreamValue.Store(downstream)
+				break
+			}
+
+			if elapsedTime > p.ReplTimeout {
+				logger.Error("error checking replication status", "err", "downstream unable to make forward progress")
+				return
+			}
+
+			logger.Debug("continuing replication check")
+			select {
+			case <-stopCh:
+				logger.Debug("canceled, stopping")
+				return
+			case <-time.After(checkInterval):
+			}
+		}
+
+		if enableWrites {
+			// TODO: enable writes, restart?
```

Comment:
With MySQL we should be able to do this without a restart.
Because we don't give out `SUPER` willy nilly it should be safe enough just to set `read_only` to `1` and things should be great. To enable writes again we can just set it back to `0`.